## Numpy

Numpy comes installed by default on Google Colaboratory. That means that you don't need to use pip to install Numpy for your work. You should know, however, that when working with Python in a "from scratch" environment, you would need to install Numpy.

Numpy was designed to make manipulating data easy. The fundamental structure that it uses is the *array*.  A Numpy array is a table of elements, all of the same type. Arrays may be one-dimensional, two-, or even higher-dimensional. 

![Numpy Arrays](assets/python4_intro_01.png)

Each dimension is called an *axis*. In a one-dimensional array, there is only a single axis; in a two- dimensional array, there are two *axes*; and in a three-dimensional array there are three axes, etc. The axes of the array are indexed by integers. We refer to individual values in the array by the index of the axis.

To work with Numpy first import the library.

In [0]:
import numpy as np

## Array Creation

There are a few ways to create an array in Numpy. First, an array can be created from a list of values.

In [0]:
# a one dimensional array
a = np.array([5, 3, 7])
print(a)

[5 3 7]


Similarly, a two-dimensional array may be created by passing a list of lists.

In [0]:
# a two dimensional array
b = np.array([[2, 4, 6], [3, 5, 7]])
print(b)

[[2 4 6]
 [3 5 7]]


Sometimes, you may want to create an array out of a sequence of numbers. For instance, to create an array of all numbers from 0 to 19:

In [0]:
c = np.arange(20)
print(c)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]


The `arange()` method can optionally take, `start`, `end`, and `step` parameters. To create an array of all even numbers from 20 to 50, for instance, we can specify a `start` of 20, `end` of 50 and `step` of 2. Note that the `end` parameter is not inclusive.

In [0]:
d = np.arange(20, 50, 2)
print(d)

[20 22 24 26 28 30 32 34 36 38 40 42 44 46 48]


In general, it's a more computationally expensive operation to resize an array than it is to create one from scratch. Because of this, sometimes it makes sense to create an array of a given size with placeholder values that you later fill in with meaningful ones. The `zero()` method provides an easy way to create an array of zeros.

In [0]:
# create an array of zeros
e = np.zeros(10)
print(e)

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [0]:
# create a two-dimensional array of ones
f = np.ones((5, 8))
print(f)

[[1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]
 [1. 1. 1. 1. 1. 1. 1. 1.]]


## Reshape

You can change the dimensions of an array with the `reshape()` method. For instance, here's what we could do if we had a one-dimensional array of 20 numbers and needed to convert this to a 4x5 array of the same numbers:

In [0]:
g = np.arange(20)
print(g)

g = g.reshape(4, 5)
print(g)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]
[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]]


## Basic Attributes

The array object has several useful attributes. For example:


*   `ndim` - the number of axes
*   `shape` - the dimensions of the array
* `size` - the number of elements in the array
* `dtype` - the data type of the elements of the array




In [0]:
# create a new array
h = np.arange(25).reshape(5, 5)

# print the number of axes in this array
print('There are {} axes in this array.'.format(h.ndim))

# print the shape of the array
print('This is a {} array'.format(h.shape))

# print the size of the array
print('This array contains {} elements'.format(h.size))

# print the data type of the elements
print('This array contains values of type {}'.format(h.dtype))

There are 2 axes in this array.
This is a (5, 5) array
This array contains 25 elements
This array contains values of type int64


## Basic Operators

A class of 30 students completed a coursework assessment worth 40% and a final exam worth 60% of the final grade. The grades for the coursework and the marks for the exam are in two arrays. Find the total for each student by adding corresponding grades from the two arrays.

To solve such a problem we need to iterate over the arrays and add corresponding values to create a thrid array. Numpy has built in operators that perform *elementwise* operations, and that's what we need here. Numpy's arithmetic operators are all elementwise.

In [0]:
# use numpy's random methods to create an array of random values
cw_marks = np.random.randint(0, 41, 30)
exam_marks = np.random.randint(0, 61, 30)

print('The coursework marks are:')
print(cw_marks)
print('The final exam marks are:')
print(exam_marks)

total = cw_marks + exam_marks

print('The total marks are:')
print(total)

The coursework marks are:
[ 1 32 31 11 11 15  0  2 26 40 22  2  6  6 33 30 39 38 32 10 19  1 31 28
  9 10 17 32 24 24]
The final exam marks are:
[21 34 52 20 18 24 18 24 36 34 19 38 42 17 40 50 34  5 23 29  7 31 31  6
 28 30 35 49  4 48]
The total marks are:
[22 66 83 31 29 39 18 26 62 74 41 40 48 23 73 80 73 43 55 39 26 32 62 34
 37 40 52 81 28 72]


Some operators on arrays are called *aggregator* functions because they aggregate the elements of the array and return a single value. For example, we can use built-in Numpy functions to find the mean, min, and max of the total grades.

In [0]:
print('The highest mark in the couse was {}'. format(np.max(total)))
print('The lowest mark in the course was {}'.format(np.min(total)))
print('The mean mark in the course was {}'.format(np.mean(total)))
print('The standard deviation of the marks in the course was {}'.format(np.std(total)))

The highest mark in the couse was 83
The lowest mark in the course was 18
The mean mark in the course was 47.63333333333333
The standard deviation of the marks in the course was 19.8099021255084


## Pandas

Pandas is an open source Python library that provides easy-to-use high performance data structures and data analysis tools. There are two main data structures provided by Pandas:

1. DataFrames
2. Series

We will spend most of our time working with DataFrames but the two structures are closely related so we'll take a brief look at Series as well.

### Series

A Pandas Series can be thought of a single column in a spreadsheet. It is a one-dimensional array of values.

In [0]:
# first import Pandas
import pandas as pd
import numpy as np

# Create a Series from an array
ages = pd.Series(np.array([20, 34, 28, 26, 39, 35, 34, 36, 38]))

# print this Series
print(ages)

0    20
1    34
2    28
3    26
4    39
5    35
6    34
7    36
8    38
dtype: int64


The numbers on the left are row index numbers that have been automatically generated. Again, imagine a spreadsheet with just a single column, even that spreadsheet will have row numbers on the left.

It's also possible to set our own string-based index.

In [0]:
age_array = np.array([50, 51, 39, 38, 124, 2013, 45, 1768, 98])
names_list = ['Sam', 'Frodo', 'Merry', 'Pippin', 'Aragorn', 'Gandalf', 'Boromir', 'Legolas', 'Gimli']

fellowship = pd.Series(age_array, index=names_list)

print(fellowship)

Sam          50
Frodo        51
Merry        39
Pippin       38
Aragorn     124
Gandalf    2013
Boromir      45
Legolas    1768
Gimli        98
dtype: int64


Alternatively, we could use a Python dictionary to create a Series.

In [0]:
# Here is a dictionary of teams and points
teams = {
    'Liverpool': 69,
    'Man City': 68,
    'Tottenham': 60,
    'Arsenal': 56,
    'Man United': 55
}

# construct a Series from the dictionary
standings = pd.Series(teams)

# Print the Series
print(standings)

Arsenal       56
Liverpool     69
Man City      68
Man United    55
Tottenham     60
dtype: int64


### Basic operations

Pandas has a few basic operations built into the Series. The full list of attributes and methods can be found in the [documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html). Here are a few examples.

In [0]:
print('The oldest age is {}'.format(fellowship.max()))
print('The youngest age is {}'.format(fellowship.min()))
print('The mean points in the league is {}'.format(standings.mean()))
print('The standard deviation of points in the league is {}'.format(standings.std()))

The oldest age is 2013
The youngest age is 38
The mean points in the league is 61.6
The standard deviation of points in the league is 6.58027355054484


The `describe()` method provides all these values with a single method call, and it can be quite helpful in analyzing central tendency.

In [0]:
standings.describe()

count     5.000000
mean     61.600000
std       6.580274
min      55.000000
25%      56.000000
50%      60.000000
75%      68.000000
max      69.000000
dtype: float64

## DataFrames

If a Series is a single column in a spreadsheet, then the DataFrame is an entire spreadsheet. It is a two- dimensional structure, with labeled axes.

![DataFrame](assets/python4_intro_02.png)


There are a number of ways to create a DataFrame, but most commonly, the data is read from some source, say a file or a database and the DataFrame is constructed from that data. We're going to explore the four most common ways to create DataFrames:

   * from csv using `read_csv()`
   * from in memory lists
   * from JSON data using `read_json()`
   * from a database using `read_sql_query()` or `read_sql_table()`
   
  
  ## From CSV

In [0]:
import pandas as pd
df = pd.read_csv('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-analytics-bootcamp/songs.csv')

The variable named `df` contain the DataFrame. Notice that we did not give any further information about the data. Pandas will attempt to infer the columns and datatypes from the data in the file. We do have some options for specifying what we expect though, but first lets explore the data that was found. There are a number of methods and attributes of the DataFrame that we can use to examine the DataFrame itself.

First, let's see the list of columns with corresponding number of values and data types.

In [0]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2229 entries, 0 to 2228
Data columns (total 8 columns):
Song Clean      2229 non-null object
ARTIST CLEAN    2229 non-null object
Release Year    1652 non-null object
COMBINED        2229 non-null object
First?          2229 non-null int64
Year?           2229 non-null int64
PlayCount       2229 non-null int64
F*G             2229 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.4+ KB


There is a lot of information in this output. The first line simply tells us that the object is a DataFrame. The second line:

```
RangeIndex: 2229 entries, 0 to 2228
```

states that an index that ranges from 0 to 2228 was created for the rows. That means that this DataFrame contains 2229 rows numbered from 0 to 2228.

Next, the column information is given. There are eight columns in total. For each column, the name of the column, the total number of non-null values and the data type of the column is given. For instance, the column named "Song Clean" has 2229 entries and is of type "object", which means its a string in this case. 

The "Release Year" column only has 1652 entries implying that some of the rows in that column contain null values. This is important to note if our analysis uses that field later.

We can list the column names.

In [0]:
df.columns

Index(['Song Clean', 'ARTIST CLEAN', 'Release Year', 'COMBINED', 'First?',
       'Year?', 'PlayCount', 'F*G'],
      dtype='object')

Or the data types.

In [0]:
df.dtypes

Song Clean      object
ARTIST CLEAN    object
Release Year    object
COMBINED        object
First?           int64
Year?            int64
PlayCount        int64
F*G              int64
dtype: object

The shape of the DataFrame is the row x column dimensions. We know that we have 2229 rows and 8 columns.

In [0]:
df.shape

(2229, 8)

## View The Data
Later we will look at various ways to select the data in a DataFrame but we will look at a simple way to take a peek at the data just to verify that we have a correct DataFrame. To view the data we can try printing the entire DataFrame. While that would work well for a small DataFrame, most of the DataFrames we will encounter in practice would probably be too large to make printing practical. For instance, printing the 2229 rows in this DataFrame would be impractical. We could just select a subset of the rows and print just those. The `head()` method prints the first 5 rows of the dataset and `tail()` prints the last 5 rows.

In [0]:
# select the first 5 rows
df.head()

Song Clean ARTIST CLEAN Release Year  \
0        Caught Up in You  .38 Special         1982   
1            Fantasy Girl  .38 Special          NaN   
2         Hold On Loosely  .38 Special         1981   
3  Rockin' Into the Night  .38 Special         1980   
4       Art For Arts Sake         10cc         1975   

                                COMBINED  First?  Year?  PlayCount  F*G  
0        Caught Up in You by .38 Special       1      1         82   82  
1            Fantasy Girl by .38 Special       1      0          3    0  
2         Hold On Loosely by .38 Special       1      1         85   85  
3  Rockin' Into the Night by .38 Special       1      1         18   18  
4              Art For Arts Sake by 10cc       1      1          1    1

In [0]:
# select the last 5 rows
df.tail()

Song Clean ARTIST CLEAN Release Year  \
2224                      She Loves My Automobile       ZZ Top          NaN   
2225                            Tube Snake Boogie       ZZ Top         1981   
2226                                         Tush       ZZ Top         1975   
2227                                   TV Dinners       ZZ Top         1983   
2228  WAITIN' FOR THE BUS/JESUS JUST LEFT CHICAGO       ZZ Top         1973   

                                               COMBINED  First?  Year?  \
2224                  She Loves My Automobile by ZZ Top       1      0   
2225                        Tube Snake Boogie by ZZ Top       1      1   
2226                                     Tush by ZZ Top       1      1   
2227                               TV Dinners by ZZ Top       1      1   
2228  WAITIN' FOR THE BUS/JESUS JUST LEFT CHICAGO by...       1      1   

      PlayCount  F*G  
2224          1    0  
2225         32   32  
2226        109  109  
2227          1    1  
2228          2    2

In [0]:
# head() and tail() takes an optional number to vary the number of rows returned.
# select the first 10 rows
df.head(10)

Song Clean   ARTIST CLEAN Release Year  \
0        Caught Up in You    .38 Special         1982   
1            Fantasy Girl    .38 Special          NaN   
2         Hold On Loosely    .38 Special         1981   
3  Rockin' Into the Night    .38 Special         1980   
4       Art For Arts Sake           10cc         1975   
5              Kryptonite   3 Doors Down         2000   
6                   Loser   3 Doors Down         2000   
7           When I'm Gone   3 Doors Down         2002   
8              What's Up?  4 Non Blondes         1992   
9              Take On Me           a-ha         1985   

                                COMBINED  First?  Year?  PlayCount  F*G  
0        Caught Up in You by .38 Special       1      1         82   82  
1            Fantasy Girl by .38 Special       1      0          3    0  
2         Hold On Loosely by .38 Special       1      1         85   85  
3  Rockin' Into the Night by .38 Special       1      1         18   18  
4              Art For Arts Sake by 10cc       1      1          1    1  
5             Kryptonite by 3 Doors Down       1      1         13   13  
6                  Loser by 3 Doors Down       1      1          1    1  
7          When I'm Gone by 3 Doors Down       1      1          6    6  
8            What's Up? by 4 Non Blondes       1      1          3    3  
9                     Take On Me by a-ha       1      1          1    1

Let us create a function that will print some of this basic information about the  DataFrame so that we can use it as we load up new DataFrames.


In [0]:
def printDataFrame(df):
  print('The shape of this DataFrame is {}'.format(df.shape))
  print('The columns in this DataFrame are {}'.format(df.columns))
  print('The info:')
  print(df.info())

In [0]:
printDataFrame(df)
df.head(2)

The shape of this DataFrame is (2229, 8)
The columns in this DataFrame are Index(['Song Clean', 'ARTIST CLEAN', 'Release Year', 'COMBINED', 'First?',
       'Year?', 'PlayCount', 'F*G'],
      dtype='object')
The info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2229 entries, 0 to 2228
Data columns (total 8 columns):
Song Clean      2229 non-null object
ARTIST CLEAN    2229 non-null object
Release Year    1652 non-null object
COMBINED        2229 non-null object
First?          2229 non-null int64
Year?           2229 non-null int64
PlayCount       2229 non-null int64
F*G             2229 non-null int64
dtypes: int64(4), object(4)
memory usage: 139.4+ KB
None


Song Clean ARTIST CLEAN Release Year  \
0  Caught Up in You  .38 Special         1982   
1      Fantasy Girl  .38 Special          NaN   

                          COMBINED  First?  Year?  PlayCount  F*G  
0  Caught Up in You by .38 Special       1      1         82   82  
1      Fantasy Girl by .38 Special       1      0          3    0

## From in memory lists

Sometimes you need to create a DataFrame from in-memory data. For instance, suppose that you had the following league table for a popular sport:

![Premier League](assets/python4_intro_03.png)

Here's how we can create a DataFrame with that data.

In [0]:
# Define a list of column labels
columns = ['club', 'MP', 'W', 'D', 'L', 'GF', 'GA', 'GD', 'Pts']

# Define rows
rows = [
    ['Man. City', 29, 23, 2, 4, 76, 20, 56, 71],
    ['Liverpool', 29, 21, 7, 1, 64, 15, 49, 70],
    ['Tottenham', 29, 20, 1, 8, 56, 30, 26, 61],
    ['Man U', 29, 17, 7, 5, 58, 38, 20, 58],
    ['Arsenal', 29, 17, 6, 6, 61, 39, 22, 57]
]

league_df = pd.DataFrame(data=rows, columns=columns)

printDataFrame(league_df)
league_df.head(2)

The shape of this DataFrame is (5, 9)
The columns in this DataFrame are Index(['club', 'MP', 'W', 'D', 'L', 'GF', 'GA', 'GD', 'Pts'], dtype='object')
The info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 9 columns):
club    5 non-null object
MP      5 non-null int64
W       5 non-null int64
D       5 non-null int64
L       5 non-null int64
GF      5 non-null int64
GA      5 non-null int64
GD      5 non-null int64
Pts     5 non-null int64
dtypes: int64(8), object(1)
memory usage: 440.0+ bytes
None


club  MP   W  D  L  GF  GA  GD  Pts
0  Man. City  29  23  2  4  76  20  56   71
1  Liverpool  29  21  7  1  64  15  49   70

Our original data here was a list of lists. If the data was all of the same type Numpy arrays could be used too. This data was represented in a row-wise manner, which means that for each row, we had a single list.

Alternatively, the data could be represented in memory with a dictionary object. In a dictionary, the data would be represented column-wise.

In [0]:
league_dict = {
    'club' : ['Man. City', 'Liverpool', 'Tottenham', 'Man U', 'Arsenal'],
    'MP' : [29, 29, 29, 29, 29],
    'W' : [23, 21, 20, 17, 17],
    'D' : [2, 7, 1, 7, 6],
    'L' : [4, 1, 8, 5, 6],
    'GF' : [76, 64, 56, 38, 39],
    'GA' : [20, 15, 30, 20, 39],
    'GD' : [56, 49, 26, 20, 22],
    'Pts' : [71, 70, 61, 58, 57]
}

league_df = pd.DataFrame(league_dict)

printDataFrame(league_df)
league_df.head(2)

The shape of this DataFrame is (5, 9)
The columns in this DataFrame are Index(['D', 'GA', 'GD', 'GF', 'L', 'MP', 'Pts', 'W', 'club'], dtype='object')
The info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 9 columns):
D       5 non-null int64
GA      5 non-null int64
GD      5 non-null int64
GF      5 non-null int64
L       5 non-null int64
MP      5 non-null int64
Pts     5 non-null int64
W       5 non-null int64
club    5 non-null object
dtypes: int64(8), object(1)
memory usage: 440.0+ bytes
None


D  GA  GD  GF  L  MP  Pts   W       club
0  2  20  56  76  4  29   71  23  Man. City
1  7  15  49  64  1  29   70  21  Liverpool

## From JSON

JSON (JavaScript Object Notation) is another ubiquitous format for data storage and distribution. JSON is easy to parse and understand, and it is well supported by most languages including Python.

The *shots.json* file contain vaccination shots given in Texas. It contains an array of data that looks like this:

```
[
  {"location":"Far South","month":"OCT","number_of_shots":"1118","program_name":"Shots for Tots","visits":"436","year":"2010"},
  {"location":"Far South","month":"NOV","number_of_shots":"1600","program_name":"Shots for Tots","visits":"879","year":"2010"},
  {"location":"Far South","month":"DEC","number_of_shots":"1379","program_name":"Shots for Tots","visits":"818","year":"2010"}
,
...
]
```

The data consists of an array of objects representing clinics. Each object has the location,  month abd year, number of shots given, and number of visitors.

In [0]:
shots_df = pd.read_json('https://tf-assets-prod.s3.amazonaws.com/tf-curric/data-analytics-bootcamp/shots.json')

printDataFrame(shots_df)
shots_df.head(2)

The shape of this DataFrame is (432, 6)
The columns in this DataFrame are Index(['location', 'month', 'number_of_shots', 'program_name', 'visits',
       'year'],
      dtype='object')
The info:
<class 'pandas.core.frame.DataFrame'>
Int64Index: 432 entries, 0 to 431
Data columns (total 6 columns):
location           432 non-null object
month              432 non-null object
number_of_shots    432 non-null int64
program_name       432 non-null object
visits             432 non-null int64
year               432 non-null int64
dtypes: int64(3), object(3)
memory usage: 23.6+ KB
None


location month  number_of_shots    program_name  visits  year
0  Far South   OCT             1118  Shots for Tots     436  2010
1  Far South   NOV             1600  Shots for Tots     879  2010

## From databases

You already have learned how to work with relational databases. Sometimes you'll want to get data from one more more databases into Pandas. To do that, we'll use what we already learned about Python + Postgres + the SQLAlchemy library.  We'll use those tools to generate a DataFrame from the results of SQL queries.

In [0]:
# import SQLAlchemy
from sqlalchemy import create_engine

# Database credentials
postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'medicalcosts'

# Create a connection
engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

# Use a database table to create a DataFrame
insurance_df = pd.read_sql_table('medicalcosts', con=engine)

# remove the connection
engine.dispose()

# Lets see some details of the DataFrame
printDataFrame(insurance_df)
insurance_df.head(2)

/usr/local/lib/python3.6/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


The shape of this DataFrame is (1338, 7)
The columns in this DataFrame are Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='object')
The info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
age         1338 non-null int64
sex         1338 non-null object
bmi         1338 non-null float64
children    1338 non-null int64
smoker      1338 non-null object
region      1338 non-null object
charges     1338 non-null float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.2+ KB
None


age     sex    bmi  children smoker     region   charges
0   19  female  27.90         0    yes  southwest  16884.90
1   18    male  33.77         1     no  southeast   1725.55

With this query, we can now think of the dataframe as an in-memory representation of an entire database table.  It's equally possible to create DataFrames from queries featuring filters or joins.

In [0]:
# Execute a query and create a DataFrame
females_df = pd.read_sql_query("SELECT age, bmi, smoker, region from medicalcosts WHERE sex = 'female'", con=engine)

# remove the connection
engine.dispose()

# Lets see some details of the DataFrame
printDataFrame(females_df)
females_df.head(2)

The shape of this DataFrame is (662, 4)
The columns in this DataFrame are Index(['age', 'bmi', 'smoker', 'region'], dtype='object')
The info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 662 entries, 0 to 661
Data columns (total 4 columns):
age       662 non-null int64
bmi       662 non-null float64
smoker    662 non-null object
region    662 non-null object
dtypes: float64(1), int64(1), object(2)
memory usage: 20.8+ KB
None


age    bmi smoker     region
0   19  27.90    yes  southwest
1   31  25.74     no  southeast